src/Workspaces/CSharp/Portable/CodeGeneration/MethodGenerator.cs

```diff
@@ -195,7 +195,7 @@ internal static class MethodGenerator
                         tokens.Add(SyntaxFactory.Token(SyntaxKind.StaticKeyword));
                     }
 
-                    if (method.IsReadOnly)
+                    if (method.IsReadOnly && (method.ContainingSymbol as INamedTypeSymbol)?.IsReadOnly != true)
```

Comment:
1. when is a method not contained by a named type?
2. I believe there's a way to ask for the containing type of a symbol

Reply:
My thought here was to guard against local functions and lambdas who are not members of their ContainingType. That might be non-applicable for MetadataAsSource specifically but could impact other areas where we use this MethodGenerator.

If this assumption is sound (maybe need to add a test for it) then I can clarify the intent in a comment here.